include/up-cpp/datamodel/validator/UMessage.h

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2024 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-FileType: SOURCE
+// SPDX-FileCopyrightText: 2024 Contributors to the Eclipse Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef UP_CPP_DATAMODEL_VALIDATOR_UMESSAGE_H
+#define UP_CPP_DATAMODEL_VALIDATOR_UMESSAGE_H
+
+#include <uprotocol/v1/umessage.pb.h>
+
+/// @brief Validators for UMessage objects.
+///
+/// See
+/// https://github.com/eclipse-uprotocol/up-spec/blob/main/basics/uattributes.adoc
+namespace uprotocol::datamodel::validator::message {
+
+/// @brief Checks if UMessage is a valid UMessage of any format.
+///
+/// A UMessage is valid if any of these are true:
+///
+///   * isValidRpcRequest()
+///   * isValidRpcResponse()
+///   * isValidPublish()
+///   * isValidNotification()
+[[nodiscard]] bool isValid(const v1::UMessage&);
+
+/// @brief Checks if common attributes for all UMessage types are valid
+///
+/// These checks must pass:
+///   * The message ID must be a valid UUID
+///   * If TTL is specified, the ID must not be expired
+///   * If Priority is specified, it is within the range of UPriority
+[[nodiscard]] bool areCommonAttributesValid(const v1::UMessage&);
+
+/// @brief Checks if UMessage is valid for invoking an RPC method
+///
+/// In addition to all common attributes being valid, these checks must pass:
+///   * Message type must be UMESSAGE_TYPE_REQUEST
+///   * Message source must pass uri::isValidRpcResponse()
+///   * Message sink must pass uri::isValidRpcMethod()
+///   * Message priority must be UPRIORITY_CS4 or higher
+///   * Message ttl must be set and greater than zero
+///   * Message must not set commstatus
+///   * Message must not set reqid
+[[nodiscard]] bool isValidRpcRequest(const v1::UMessage&);
```

Comment:
Might be nice to have a return type for these APIs which can state why it was invalid.

The [`up-rust` implementation](https://github.com/eclipse-uprotocol/up-rust/blob/04c9518bb451bcab233be3f762ec68e99ea67494/src/uattributes/uattributesvalidator.rs#L36) returns `Result<(), UAttributesError>` which allows us to note (at least in String form) why it was failed.

Reply:
Maybe something like
```
enum class Reason {
    WRONG_MESSAGE_TYPE,
    TTL_NOT_SET,
    BAD_UUID,
    EXPIRED_ID,
    ...
};

[[nodiscard]] std::tuple<bool, std::optional<Reason>> isValidRpcRequest(const v1::UMessage&);
```

This way the check can be done something like:
```
auto [valid_request, maybe_invalid_reason] = isValidRpcRequest(some_message);
if (!valid_request) {
    // check the reason and do something / throw an exception
}
```

Reply:
Yeah, this would be even better IMHO than what we currently have in up-rust, with a more concrete set of reasons.